Fix buffer issues with async IPC; add tests

fselmo opened the pull request:
### What was wrong?

Closes #3485

### How was it fixed?

- Move to a `readline()` approach on the stream reader for IPC. The older implementation for the original IPC would read in chunks but this isn't ideal and, it turns out, opting for readline while allowing control of the buffer limit lends itself quite nicely to plucking well-formed responses from the socket.

### Todo:

- [x] Clean up commit history / elaborate on commit message
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![ezgif com-animated-gif-maker](https://github.com/user-attachments/assets/9009755f-b4bc-4afd-a2b9-eab4f8a52273)

fselmo committed:
Fix buffer issues with async IPC; add tests:

- Move to a readline() approach on the stream
  reader for IPC. The older implementation for
  the original IPC would read in chunks but this
  isn't ideal and, it turns out, opting for
  readline while allowing control of the buffer
  limit lends itself quite nicely to plucking
  well-formed responses from the socket.

raxhvl commented:
I wonder if chunks can arrive in arbitrary order. Also [this thread discusses](https://groups.google.com/g/nodejs/c/3LHxMZRev1E?pli=1) possible solutions.

fselmo committed:
Add documentation around read buffer; Raise clear exception:

- Document ``read_buffer_limit`` on the ``AsyncIPCProvider`` class.

- Add ``ReadBufferLimitReached`` exception to be raised when the read buffer
  limit is reached, prompting the user to increase the limit.

- Add a base ``PersistentConnectionError`` exception class for persistent
  connection errors to inherit from.

fselmo commented:
@raxhvl thanks for reporting. I believe this will be a more straightforward approach that can give control to the user to configure the read buffer limit if needs to be even larger than the default of 20MB. This should resolve your issues and I added a test to test this limit.

kclowes commented on web3/providers/persistent/async_ipc.py:
```diff
@@ -101,18 +97,20 @@ async def socket_send(self, request_data: bytes) -> None:
-                self._raw_message += to_text(await self._reader.read(4096)).lstrip()
+        try:
+            data = await self._reader.readline()
+        except ValueError as e:
```
Do we want to translate this into a `Web3ValueError` or just raise `ValueError` here?

fselmo replied:
Do you mean to also extend from `Web3ValueError` for the `ReadBufferLimitReached` exception?

fselmo replied:
^ I think that makes sense. I can only see it helping in context and in handling the error as well.

kclowes commented on tests/core/providers/test_async_ipc_provider.py:
```diff
@@ -264,3 +298,43 @@ async def test_async_iterator_pattern_exception_handling_for_subscriptions(
+            pathlib.Path(jsonrpc_ipc_pipe_path), read_buffer_limit=TWENTY_MB + 1024
+        )
+    ) as w3:
+        await w3.provider.make_request("method", [])
```
Do we want to assert anything about the response here like we do on lines 309-311?

kclowes reviewed (approved):
I left a couple questions to take or leave, but generally looks good to me! Much cleaner! 🧹

fselmo committed:
Changes from comments on PR #3492:

- Also inherit from ``Web3ValueError`` for ``ReadBufferLimitReached``.

- Properly assert the response for configured read limit test.
